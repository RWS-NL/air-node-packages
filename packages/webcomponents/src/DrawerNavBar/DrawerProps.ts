import {
  AppBarProps as MAppBarProps,
  BoxProps as MBoxProps,
  DividerProps as MDividerProps,
  DrawerProps as MDrawerProps,
  ListItemProps as MListItemProps,
  TabsProps as MTabsProps,
  ToolbarProps as MToolbarProps,
  TooltipProps as MTooltipProps
} from '@material-ui/core';
import { ChangeEvent, ReactNode } from 'react';
import { LinkTabProps } from '../LinkTab';

export interface DrawerProps {
  /**
   * Whether the navigation drawer is open or not
   */
  open: boolean;
  /**
   * The current width of the navigation drawer
   * @remark only applies when the drawer is expanded
   */
  width: number;
  /**
   * Function to trigger when toggling the navigation drawer
   * @example
   * ```ts
   * const handleToggleDrawer = () => setOpen(!open)
   * ```
   */
  toggleDrawer: () => void;
}

export interface DrawerNavBarProps {
  /** Props for the Tree drawer */
  treeDrawer: TreeDrawerProps;
  /** Props for the navigation drawer */
  navigationDrawer: NavigationDrawerProps;
  /** The tabs to render in the navigation bar */
  tabs: LinkTabProps[];
  /**
   *  The currently active tab
   *  Preferably managed in local state
   */
  activeTab: number;
  /** Function to trigger when changing tabs */
  onTabChange(_event: ChangeEvent<unknown>, newValue: number): void;

  /** Additional props to pass to the AppBar component */
  AppBarProps?: MAppBarProps;
  /** Additional props to pass to the Toolbar component */
  ToolbarProps?: MToolbarProps;
  /** Additional props to pass to the Tabs component */
  TabsProps?: MTabsProps;

  /** Additional props to pass to the NavigationHeader component */
  NavigationHeaderProps?: MAppBarProps;
  /** Additional props to pass to the NavigationDrawer component */
  NavigationDrawerProps?: MDrawerProps;
  /** Additional props to pass to each NavigationDrawerItem component */
  NavigationDrawerItemProps?: MListItemProps;
  /** Additional props to pass to the TreeDrawer component */
  TreeDrawerProps?: MDrawerProps;
  /** Additional props to pass to the ResizableDrawer component */
  ResizableDividerProps?: MDividerProps;
  /** Additional props to pass to the Content component */
  ContentBoxProps?: MBoxProps;
}

export interface NavigationDrawerProps extends DrawerProps {
  /**
   * The items to show in the navigation drawer
   */
  items: DrawerItem[];
  /** Additional props to pass to each NavigationDrawerItem component */
  NavigationDrawerItemProps?: MListItemProps;
}

export interface DropdownValue {
  value: string;
  label: string;
}

export interface TreeDrawerProps extends DrawerProps {
  /**
   * The minimum width of the navigation drawer
   * @remark only applies when the drawer is expanded
   * @remark The width of the drawer cannot go lower than the minWidth.
   */
  minWidth: number;
  /** The content of the tree */
  content: ReactNode;
  /** The text to be displayed in the Tooltip when hovering over the info icon */
  tooltipText: MTooltipProps['title'];
  /**
   * Event to trigger when dragging the right-most border of the drawer
   * @example
   * // Using lodash.throttle to throttle the event for performance
   *
   * const handleDragTreeBorder = throttle(
   *   (_, value: unknown) => {
   *     setWidth(value as number);
   *   },
   *   10,
   *   { leading: true, trailing: true }
   * );
   */
  onBorderDrag?: (
    event: React.MouseEvent<HTMLDivElement, MouseEvent> | React.TouchEvent<HTMLDivElement>,
    ...args: unknown[]
  ) => void;
  /**
   * A list of values that are selectable in the dropdown menu
   * @example
   * ```ts
   * const boomTypes: DropdownValue[] = [
   *   { value: 'Option One', label: 'Option One' },
   *   { value: 'Option Two', label: 'Option Two' },
   *   { value: 'Option Three', label: 'Option Three' },
   *   { value: 'Option Four', label: 'Option Four' },
   *   { value: 'Option Five', label: 'Option Five' },
   *   { value: 'Option Six', label: 'Option Six' }
   * ];
   * ```
   */
  dropdownValues: DropdownValue[];
  /** The current value in the dropdown menu */
  currentDropdownValue: string;
  /**
   * Event to trigger when changing the value in the dropdown menu atop the drawer
   * @example
   * const handleChange = (event: React.ChangeEvent<{ name?: string; value: unknown }>) => setDropdownValue(event.target.value);
   */
  hanbdleDropdownChange: (event: React.ChangeEvent<{ name?: string; value: unknown }>) => void;
  /** Whether the details pane should be open */
  detailsPaneOpen: boolean;
  /**
   * The content to display in the details pane popper
   * @remark This gets wrapped in a Paper element so it falls over the rest of the content
   */
  detailsPaneContent: ReactNode;
  /** Props applied to the `Box` element containing the content of the tree */
  TreeContentBoxProps?: MBoxProps;
}

/**
 * Describes a single icon in the drawer
 * @example
 * ```jsx
 * const navigationDrawerItems: DrawerItem[] = [
 *   {
 *     icon: (
 *       <ListItemIcon>
 *         <SendIcon />
 *       </ListItemIcon>
 *     ),
 *     label: (
 *       <ListItemText
 *         primary="Send"
 *         primaryTypographyProps={{ variant: "body2", color: "primary" }}
 *       />
 *     ),
 *
 *   },
 * ];
 * ```
 */
export interface DrawerItem {
  /**
   * The icon to show for this item
   * @remark Should use `<ListItemIcon>` from `@material-ui/core/ListItemIcon`
   * @example
   * ```jsx
   * <ListItemIcon>
   *   <SendIcon />
   * </ListItemIcon>
   * ```
   */
  icon: ReactNode;
  /**
   * The text label to show next to the icon when the drawer is expanded
   * @remark Should use `<ListItemText>` from `@material-ui/core/ListItemText`
   * @example
   * ```jsx
   * <ListItemText
   *   primary="Send"
   *   primaryTypographyProps={{ variant: "body2", color: "primary" }}
   * />
   * ```
   */
  label: ReactNode;
  /**
   * Some event to trigger when clicking this `<ListItem>` entry
   * @example
   * ```ts
   * () => history.push('/projects')
   * ```
   */
  onClick?: (event: React.MouseEvent<HTMLLIElement & HTMLDivElement, MouseEvent>) => void;
}

export type TreeAndNavProps = Pick<DrawerNavBarProps, 'treeDrawer' | 'navigationDrawer'>;

export interface TreeDrawerComponentProps extends TreeAndNavProps {
  onMouseDown: (event: React.MouseEvent<HTMLHRElement>) => void;
  onTouchStart: (event: React.TouchEvent<HTMLHRElement>) => void;
  /** Additional props to pass to the ResizableDrawer component */
  ResizableDividerProps?: MDividerProps;
}
