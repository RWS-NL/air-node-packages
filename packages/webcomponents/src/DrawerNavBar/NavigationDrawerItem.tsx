import { ListItem } from '@material-ui/core';
import React, { memo } from 'react';
import { DrawerItem } from './DrawerProps';

export const NavigationDrawerItem = memo(
  ({ icon, label, onClick, ...props }: DrawerItem) => (
    <ListItem button onClick={onClick} {...props}>
      {icon}
      {label}
    </ListItem>
  ),
  () => true
);
