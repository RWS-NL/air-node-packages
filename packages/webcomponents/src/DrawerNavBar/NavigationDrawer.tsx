import Divider from '@material-ui/core/Divider';
import Drawer from '@material-ui/core/Drawer';
import Grid from '@material-ui/core/Grid';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemIcon from '@material-ui/core/ListItemIcon';
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles';
import CloseIcon from '@material-ui/icons/Close';
import MenuIcon from '@material-ui/icons/Menu';
import clsx from 'clsx';
import { stringify } from 'flatted';
import React, { memo } from 'react';
import { Else, If, Then } from 'react-if';
import { NavigationDrawerProps } from './DrawerProps';
import { NavigationDrawerItem } from './NavigationDrawerItem';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    navigationDrawer: {
      flexShrink: 0,
      whiteSpace: 'nowrap'
    },
    navigationDrawerOpen: {
      width: ({ width }: Pick<NavigationDrawerProps, 'width'>) => width,
      transition: theme.transitions.create('width', {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.enteringScreen
      })
    },
    navigationDrawerClose: {
      transition: theme.transitions.create('width', {
        easing: theme.transitions.easing.sharp,
        duration: theme.transitions.duration.leavingScreen
      }),
      overflowX: 'hidden',
      width: theme.spacing(7) + 1,
      [theme.breakpoints.up('sm')]: {
        width: theme.spacing(9) + 1
      }
    },
    drawerPaper: {
      marginTop: theme.spacing(6),
      [theme.breakpoints.down('sm')]: {
        marginTop: theme.spacing(6)
      }
    },
    divider: {
      backgroundColor: '#CCE1EB' // Dark Blue 20
    },
    closeListItem: {
      justifyContent: 'flex-end'
    },
    navgationDrawerCloseIcon: {
      marginRight: theme.spacing(2)
    }
  })
);

export const NavigationDrawer = memo(
  ({ open, items, width, toggleDrawer, NavigationDrawerItemProps, ...props }: NavigationDrawerProps) => {
    const classes = useStyles({ width });

    return (
      <Drawer
        {...props}
        variant='permanent'
        className={clsx(classes.navigationDrawer, {
          [classes.navigationDrawerOpen]: open,
          [classes.navigationDrawerClose]: !open
        })}
        classes={{
          paper: clsx(classes.drawerPaper, {
            [classes.navigationDrawerOpen]: open,
            [classes.navigationDrawerClose]: !open
          })
        }}
      >
        <List>
          <ListItem
            button
            disableGutters={open}
            onClick={toggleDrawer}
            classes={{
              root: clsx({ [classes.closeListItem]: open })
            }}
          >
            <Grid
              container
              direction='row'
              justify={open ? 'flex-end' : 'flex-start'}
              alignItems='center'
              wrap='nowrap'
              alignContent='stretch'
            >
              <Grid
                item
                xs={1}
                classes={{
                  root: clsx({
                    [classes.navgationDrawerCloseIcon]: open
                  })
                }}
              >
                <ListItemIcon>
                  <If condition={open}>
                    <Then>
                      <CloseIcon color='primary' />
                    </Then>
                    <Else>
                      <MenuIcon color='primary' />
                    </Else>
                  </If>
                </ListItemIcon>
              </Grid>
            </Grid>
          </ListItem>
        </List>

        <Divider classes={{ root: classes.divider }} />

        <List>
          {items.map(({ icon, label, onClick }, index) => (
            <NavigationDrawerItem
              key={index}
              icon={icon}
              label={label}
              onClick={onClick}
              {...NavigationDrawerItemProps}
            />
          ))}
        </List>
      </Drawer>
    );
  },
  (prevProps, nextProps) => {
    if (prevProps.open !== nextProps.open) return false;
    if (prevProps.items.length !== nextProps.items.length) return false;
    if (stringify(prevProps.items) !== stringify(nextProps.items)) return false;
    return true;
  }
);
