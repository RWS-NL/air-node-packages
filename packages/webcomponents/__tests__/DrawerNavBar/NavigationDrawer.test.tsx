import { NavigationDrawerProps } from '@DrawerNavBar/DrawerProps';
import { NavigationDrawer } from '@DrawerNavBar/NavigationDrawer';
import { NavigationDrawerItem } from '@DrawerNavBar/NavigationDrawerItem';
import { navigationDrawerProps } from '@Resource/commonDrawerNavBarProps';
import { shallow, ShallowWrapper } from 'enzyme';
import React from 'react';

describe('NavigationDrawer', () => {
  describe('Component Testing', () => {
    let navDrawer: ShallowWrapper<NavigationDrawerProps>;

    beforeAll(() => {
      navDrawer = shallow(<NavigationDrawer {...navigationDrawerProps} />);
    });

    test('GIVEN items THEN renders ListItem for each', () => {
      expect(navDrawer.find(NavigationDrawerItem)).toHaveLength(navigationDrawerProps.items.length);
    });
  });

  describe('Snapshot Testing', () => {
    test('Required Props', () => {
      const navigationDrawer = shallow(
        <NavigationDrawer
          open={navigationDrawerProps.open}
          items={navigationDrawerProps.items}
          width={navigationDrawerProps.width}
          toggleDrawer={navigationDrawerProps.toggleDrawer}
        />
      );
      expect(navigationDrawer).toMatchSnapshot();
    });

    test('Optional Props', () => {
      const navigationDrawer = shallow(<NavigationDrawer {...navigationDrawerProps} />);
      expect(navigationDrawer).toMatchSnapshot();
    });
  });
});
