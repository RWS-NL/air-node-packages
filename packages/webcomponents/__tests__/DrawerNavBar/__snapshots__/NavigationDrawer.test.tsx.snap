// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavigationDrawer Snapshot Testing Optional Props 1`] = `
<WithStyles(ForwardRef(Drawer))
  className="makeStyles-navigationDrawer-1 makeStyles-navigationDrawerOpen-2 makeStyles-navigationDrawerOpen-10"
  classes={
    Object {
      "paper": "makeStyles-drawerPaper-4 makeStyles-navigationDrawerOpen-2 makeStyles-navigationDrawerOpen-10",
    }
  }
  variant="permanent"
>
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      classes={
        Object {
          "root": "makeStyles-closeListItem-6",
        }
      }
      disableGutters={true}
      onClick={[MockFunction]}
    >
      <WithStyles(ForwardRef(Grid))
        alignContent="stretch"
        alignItems="center"
        container={true}
        direction="row"
        justify="flex-end"
        wrap="nowrap"
      >
        <WithStyles(ForwardRef(Grid))
          classes={
            Object {
              "root": "makeStyles-navgationDrawerCloseIcon-7",
            }
          }
          item={true}
          xs={1}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <If
              condition={true}
            >
              <Then>
                <CloseIcon
                  color="primary"
                />
              </Then>
              <Else>
                <MenuIcon
                  color="primary"
                />
              </Else>
            </If>
          </WithStyles(ForwardRef(ListItemIcon))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Divider))
    classes={
      Object {
        "root": "makeStyles-divider-5",
      }
    }
  />
  <WithStyles(ForwardRef(List))>
    <Memo()
      icon={
        <WithStyles(ForwardRef(ListItemIcon))>
          <Memo(SendIcon) />
        </WithStyles(ForwardRef(ListItemIcon))>
      }
      key="0"
      label={
        <WithStyles(ForwardRef(ListItemText))
          primary="Send"
          primaryTypographyProps={
            Object {
              "color": "primary",
              "variant": "body2",
            }
          }
        />
      }
    />
    <Memo()
      icon={
        <WithStyles(ForwardRef(ListItemIcon))>
          <Memo(StarBorderIcon) />
        </WithStyles(ForwardRef(ListItemIcon))>
      }
      key="1"
      label={
        <WithStyles(ForwardRef(ListItemText))
          primary="Star"
          primaryTypographyProps={
            Object {
              "color": "primary",
              "variant": "body2",
            }
          }
        />
      }
    />
  </WithStyles(ForwardRef(List))>
</WithStyles(ForwardRef(Drawer))>
`;

exports[`NavigationDrawer Snapshot Testing Required Props 1`] = `
<WithStyles(ForwardRef(Drawer))
  className="makeStyles-navigationDrawer-1 makeStyles-navigationDrawerOpen-2 makeStyles-navigationDrawerOpen-9"
  classes={
    Object {
      "paper": "makeStyles-drawerPaper-4 makeStyles-navigationDrawerOpen-2 makeStyles-navigationDrawerOpen-9",
    }
  }
  variant="permanent"
>
  <WithStyles(ForwardRef(List))>
    <WithStyles(ForwardRef(ListItem))
      button={true}
      classes={
        Object {
          "root": "makeStyles-closeListItem-6",
        }
      }
      disableGutters={true}
      onClick={[MockFunction]}
    >
      <WithStyles(ForwardRef(Grid))
        alignContent="stretch"
        alignItems="center"
        container={true}
        direction="row"
        justify="flex-end"
        wrap="nowrap"
      >
        <WithStyles(ForwardRef(Grid))
          classes={
            Object {
              "root": "makeStyles-navgationDrawerCloseIcon-7",
            }
          }
          item={true}
          xs={1}
        >
          <WithStyles(ForwardRef(ListItemIcon))>
            <If
              condition={true}
            >
              <Then>
                <CloseIcon
                  color="primary"
                />
              </Then>
              <Else>
                <MenuIcon
                  color="primary"
                />
              </Else>
            </If>
          </WithStyles(ForwardRef(ListItemIcon))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(ListItem))>
  </WithStyles(ForwardRef(List))>
  <WithStyles(ForwardRef(Divider))
    classes={
      Object {
        "root": "makeStyles-divider-5",
      }
    }
  />
  <WithStyles(ForwardRef(List))>
    <Memo()
      icon={
        <WithStyles(ForwardRef(ListItemIcon))>
          <Memo(SendIcon) />
        </WithStyles(ForwardRef(ListItemIcon))>
      }
      key="0"
      label={
        <WithStyles(ForwardRef(ListItemText))
          primary="Send"
          primaryTypographyProps={
            Object {
              "color": "primary",
              "variant": "body2",
            }
          }
        />
      }
    />
    <Memo()
      icon={
        <WithStyles(ForwardRef(ListItemIcon))>
          <Memo(StarBorderIcon) />
        </WithStyles(ForwardRef(ListItemIcon))>
      }
      key="1"
      label={
        <WithStyles(ForwardRef(ListItemText))
          primary="Star"
          primaryTypographyProps={
            Object {
              "color": "primary",
              "variant": "body2",
            }
          }
        />
      }
    />
  </WithStyles(ForwardRef(List))>
</WithStyles(ForwardRef(Drawer))>
`;
